### DISCUSSION TITLE
Paymaster Compatibility and Visibility

### DISCUSSION THREAD

**Discussion by damien75** [zkSync-Community-Hub/zksync-developers#16] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

While going through the [hello world tutorial](https://era.zksync.io/docs/dev/building-on-zksync/hello-world.html#paying-fees-using-testnet-paymaster), and more specifically the section around using the paymaster, I would like to make sure that I have indeed used a token instead of ETH to pay.

### What is the specific issue or error you're encountering?

Comparing [transaction 1](https://goerli.explorer.zksync.io/tx/0x1c1a72f56aac4a30bc971d85a3e8bf17381955c40ca76e60a82015e80c5f369a) where I used ETH and not the paymaster to pay vs [transaction 2](https://goerli.explorer.zksync.io/tx/0x757df87082bda5e67e239682cb071689e47019a8556c7529eb4b63040e644fb8) where I selected DAI to pay for the fees, I can't see in the explorer whether that's actually what happened.
1. How can I confirm this?
2. I am not sure whether the hello world contract supports account abstraction where we call the contract that can itself call the paymaster. Or can any smart contract on zksync leverage any custom paymaster by default? This doesn't sound right especially if there are custom paymasters I would assume that those would need to be included as supported in the smart contract itself?
3. For that [transaction 2](https://goerli.explorer.zksync.io/tx/0x757df87082bda5e67e239682cb071689e47019a8556c7529eb4b63040e644fb8) , was any ETH used to pay for the fee, was it a combination of ETH and DAI, or was it just DAI? The explorer doesn't seem to make that clear

### Can you share the error messages or logs you're receiving, if any?

tx 1 paying with ETH: https://goerli.explorer.zksync.io/tx/0x1c1a72f56aac4a30bc971d85a3e8bf17381955c40ca76e60a82015e80c5f369a

tx 2 paying with DAI:
https://goerli.explorer.zksync.io/tx/0x757df87082bda5e67e239682cb071689e47019a8556c7529eb4b63040e644fb8

### Have you made any recent changes to the contract before encountering this issue?

I did not add the [AA signature checker](https://era.zksync.io/docs/dev/developer-guides/aa.html#aa-signature-checker) in my Greeter contract as I wasn't sure whether this was necessary, although mentioned in the tutorial

### Are there any external libraries or contracts that your contract interacts with?

no

### Can you provide the relevant portions of your contract code where the issue is occurring?

The [contract is verified on the explorer](https://goerli.explorer.zksync.io/address/0xcb847Ff1263311EF264184b90DB53480cDB002b2)

### Have you tried to isolate the problem, and if so, what were the results?

Next steps would be for me to setup a custom paymaster. I would probably need to try on a local setup. Will try following [this guide](https://era.zksync.io/docs/tools/hardhat/testing.html)

### What steps have you already taken to try to resolve the issue?

Looking at the [verified contract on the explorer](https://goerli.explorer.zksync.io/address/0xcb847Ff1263311EF264184b90DB53480cDB002b2), it looks like when we use a paymaster, we actually send ETH to the contract which is probably used to pay for the swap? I am not sure if that ETH can then be part of the refund that's sent back?

**Comment by uF4No** [comment, marked as answer]
Hi @damien75 

In the Quickstart tutorial, the paymaster integration is done in the frontend. If you select a token different from ETH, the frontend will add additional parameters to the transaction in the `getOverrides` method ([here](https://era.zksync.io/docs/dev/building-on-zksync/hello-world.html#paying-fees-using-testnet-paymaster)).

It's true that currently in the explorer this is not easy to spot but, if you open the Fee Details, you'll see that both transactions have a refunds for unused gas, and in the first one, the refund was sent to the address that sent the transaction ( 0xE5aEca5fd7EBaACa75d0B8c56f91fE98768826A8) and in the second one, the refund was sent to 0x8f0eA1312dA29f17eABEB2F484Fd3C112cCCdd63, which is the testnet paymaster address. This means you as a user paid the gas fee in DAI but behind the scenes, the paymaster actually paid the gas fee in ETH and received a refund from the protocol.

The easiest way for you to also confirm this, is to check you ETH balance before and after sending the transaction, or even sending the transaction from an account that has 0ETH.

**Comment by mattbrathwaite** [comment]
To confirm whether a token was used to pay for transaction fees instead of ETH, you can inspect the transaction details on the zkSync Explorer. Look for the "Fee" section, which should indicate the token used for fee payment. Additionally, consider implementing a custom paymaster to control fee payment behavior more precisely.

**Comment by fullertasha** [comment]
To confirm whether DAI or ETH was used to pay for transaction fees in zkSync Explorer:
1. Check the token used for the fee payment in the transaction details.
2. Verify if the contract supports account abstraction for interacting with a custom paymaster.
3. Review the contract code and explore the possibility of including AA signature checker if necessary.

**Comment by unticcutecerin** [comment]
The user is trying to confirm whether they used a token instead of ETH to pay for fees while following the hello world tutorial. They are comparing two transactions, one where they used ETH and another where they selected DAI to pay for fees, but they cannot see in the explorer whether DAI was used for the second transaction. The user is also unsure if the hello world contract supports account abstraction and if any smart contract on zksync can leverage any custom paymaster by default. Without further information or access to the code repository, it is difficult to provide a specific solution. However, it may be helpful for the user to review the documentation on zksync's paymaster feature and account abstraction to better understand how they can confirm whether they used a token instead of ETH to pay for fees. Additionally, it may be useful to check if the explorer provides any additional information on the transactions, such as the token used for fees or the breakdown of fees between ETH and DAI.
